Fixed a critical application non-exit bug

SteveBriley opened the pull request:
- Fixed a CRITICAL bug where the application wouldn't quit due to a misalignment of the wait groups.
- Removed the use of channels, as it wasn't actually required, and was potentially causing minimal channel sending/receiving contention.
- Refactored the goroutine spawning code.
- Fixed a README typo.

0xCC00FFEE commented:
Thank you Steve for the PR.
Just couple of questions before merging:
- What is the misalignment bug that you found in the wait groups?
- What do you technically mean by the contention caused by the usage of channels? Even if minimal, could it cause at some point a race condition?

SteveBriley commented:
> Thank you Steve for the PR.
> Just couple of questions before merging:
> 
> * What is the misalignment bug that you found in the wait groups?
> * What do you technically mean by the contention caused by the usage of channels? Even if minimal, could it cause at some point a race condition?

1) When I ran the application, using a single port to a single port, and even with 1-65535, it would stay open and wouldn't exit. 

With wait groups, if we add a wait at some point, but don't close a wait, and we have the application wait until all the waits are closed, the wait counter never reaches zero to permit the exit to occur.

So, I reviewed the code, eliminated the channels entirely as they weren't really needed for what was being done, and simplified the wait group adds, and the problem went away.

2) I don't recall seeing any race conditions in the application. The reason I mentioned channel contention, was because information was being passed via channels when it wasn't necessary, and as they were unbuffered channels by default, that means they can only pass one message at a time. When I removed the channels, that eliminated any potential delays or overhead associated with passing data over the channels.
